project/ian-chiu/README.md

```diff
@@ -0,0 +1,185 @@
+
+Use parallel computing techinques, like multithreading and SIMD operations, to increase computation speed of calculating all cell states in the next frame under a rule. 
+
+If a grid is 100x100x100, we have to calculate next states of 1000000 cells according to their neighbors every frame, which needs a lot of computation.
+
+#### Greedy Meshing Algorithm
+
+The idea of the algorithm is combining adjacent blocks to reduce the part count while visually keeping everything the same, which is useful to reduce draw calls. 
+
+For a 100x100x100 grid, we have to draw maximum 1000000 cubes if all cells are not empty. If the number of draw calls is too large, the communication between cpu and gpu will be overwhelming, which can greatly reduce the performance of rendering.
+
+![](https://i.imgur.com/SCIFdNj.png)
+
+## Prospective Users
+
+### Biology
+
+Several biological processes and systems are simulated using Celluar automations, including the patterns of some seashells, moving wave patterns on the skin of cephalopods and the behaviors of brain neurons.
+
+### Physics
+
+Celluar automata are used to simulate and study physical phenomena, such as gas and fluid dynamics.
+
+### Cryptography
+
+Celluar automata are proposed for use in public key cryptography. They can be utilized to construct pseudorandom number generators, and to design error correction codes.
+
+### Generative Art and Music
+
+Cellular automata have been used in generative music and evolutionary music composition and procedural terrain generation in video games.
+
+*Sources*:
+* Tech Target: https://www.techtarget.com/searchenterprisedesktop/definition/cellular-automaton
+* Wikipedia:
+https://en.wikipedia.org/wiki/Cellular_automaton#Applications
+
+## System Architecture
+
+### Features
+
+The software should have following features:
+
+1. Simulation operations: randomize, run, pause, clear
+2. Simulation settings: randomize density and range, grid size
+3. Some cell rules can be used by default
+4. Users can add their own cells rules
+5. Basic camera control: Orbiting and Zooming
+
+### Input / Output
+
+* Input: GUI, keyboard, mouse
+* Ouput: application window
+
+### Third Party Library
+
+- raylib-py or pygame: A simple video games programming library for dealing with windows, keyboard/mouse input, graphics and GUI.
+
+## API Description
+
+### Python
+
+* Rule: 
+    * apply: apply the cell rule
+    * count_neighbors: get number of alive neighbors according to the rule
+* CameraController:
+    * update: update camera view and projection matrix according to the control input
+* Application:
+    * run: control the render loop and the all update for the next frame
+    * randomize: randomize all cells in a grid
+    * clear: empty all cells in a grid
+
+### C++
+
+* calculateNextState: calculate all cells' states of the next frame. Should use multithreading and SIMD or any parallel computing techniques to increase computation speed.
```

Comment:
If multi-threading is implemented, it is correct to put it in the C++ level.

Reply:
Right, I think so too.